cfg/optimize.go

```diff
@@ -0,0 +1,296 @@
+
+var (
+	// DefaultOptimizationConfig provides a default configuration for optimizations.
+	DefaultOptimizationConfig = OptimizationConfig{
+		FlagOverrideSets: []FlagOverrideSet{
+			{
+				Name: "high-performance",
+				Overrides: map[string]FlagOverride{
+					"write.enable-streaming-writes":         {NewValue: true},
+					"metadata-cache.negative-ttl-secs":      {NewValue: 0},
+					"metadata-cache.ttl-secs":               {NewValue: -1},
+					"metadata-cache.stat-cache-max-size-mb": {NewValue: 1024},
+					"metadata-cache.type-cache-max-size-mb": {NewValue: 128},
+					"implicit-dirs":                         {NewValue: true},
+					"file-system.rename-dir-limit":          {NewValue: 200000},
+				},
+			},
+		},
+		MachineTypes: []MachineType{
+			{
+				Names: []string{
+					"a2-megagpu-16g", "a2-ultragpu-8g", "a3-edgegpu-8g", "a3-highgpu-8g", "a3-megagpu-8g", "a3-ultragpu-8g", "a4-highgpu-8g-lowmem",
+					"ct5l-hightpu-8t", "ct5lp-hightpu-8t", "ct5p-hightpu-4t", "ct5p-hightpu-4t-tpu", "ct6e-standard-4t", "ct6e-standard-4t-tpu", "ct6e-standard-8t", "ct6e-standard-8t-tpu"},
+				FlagOverrideSetName: "high-performance",
+			},
+			// Add more machine types here as needed.
+		},
+	}
+
+	// metadataEndpoints are the endpoints to try for fetching metadata.
+	// Use an array to make provision for https endpoint in the future: https://cloud.google.com/compute/docs/metadata/querying-metadata#metadata_server_endpoints
+	metadataEndpoints = []string{
+		"http://metadata.google.internal/computeMetadata/v1/instance/machine-type",
+	}
+)
+
+////////////////////////////////////////////////////////////////////////
+// Helper Functions
+////////////////////////////////////////////////////////////////////////
+
+// getMachineType fetches the machine type from the metadata server.
+func getMachineType(isSet isValueSet) (string, error) {
+	// Check if the machine-type flag is set and not empty.
+	if isSet.IsSet("machine-type") {
+		machineType := isSet.GetString("machine-type")
+		if machineType != "" {
+			return machineType, nil
+		}
+	}
+	client := http.Client{Timeout: httpTimeout}
+	for retry := 0; retry < maxRetries; retry++ {
+		for _, endpoint := range metadataEndpoints {
+
+			req, err := http.NewRequest(http.MethodGet, endpoint, nil)
+			if err != nil {
+				return "", fmt.Errorf("failed to create request for %s: %w", endpoint, err)
+			}
+			req.Header.Add("Metadata-Flavor", "Google")
+
+			resp, err := client.Do(req)
+			if err != nil {
+				continue // Try the next endpoint.
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode != http.StatusOK {
+				continue
+			}
+
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return "", fmt.Errorf("failed to read response body from %s: %w", endpoint, err)
+			}
+
+			machineType := string(body)
+			parts := strings.Split(machineType, "/")
+			if len(parts) > 0 {
+				return parts[len(parts)-1], nil
+			}
+			return machineType, nil
```

Comment:
This seems equivalent to:
```suggestion
			return parts[len(parts)-1], nil
```

Reply:
This is to avoid crash in case of machineType turned out to be empty.